compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt

```diff
@@ -138,3 +137,12 @@ fun JvmBackendContext.createJvmIrBuilder(
 
 fun IrDeclaration.isInCurrentModule(): Boolean =
     getPackageFragment() is IrFile
+
+// Determine if the IrExpression is smartcast, and if so, if it is cast from higher than nullable target types.
+// This is needed to pinpoint exceptional treatment of IEEE754 floating point comparisons, where proper IEEE
+// comparisons are used "if values are statically known to be of primitive numeric types", taken to mean as
+// "not learned through smartcasting".
+fun IrExpression.isSmartcastFromHigherThanNullable(context: JvmBackendContext) =
+    (this as? IrTypeOperatorCall)?.let {
```

Comment:
A bit cleaner would be
```
    this is IrTypeOperatorCall &&
            operator == IrTypeOperator.IMPLICIT_CAST &&
            !argument.type.isSubtypeOf(type.makeNullable(), context.irBuiltIns)
```

Also, I'd suggest expanding the last condition into:
```
            !argument.type.isNullable() && !argument.type.isSubtypeOf(type, context.irBuiltIns)
```
since it seems a bit easier to understand and checks the same (please correct me if I'm wrong)

Reply:
If we're casting from Any? to Float?,  that is indeed "too high" a smart-cast: Any? is _not_ a subtype of Float?.makeNullable().

But the new definition OK's it as Any? is nullable.

Reply:
You're right, sorry. I intended the following:
```
            (!argument.type.isNullable() || !argument.type.isSubtypeOf(type, context.irBuiltIns))
```

But then it's not much easier than the original anymore, so never mind. :)